okhttp/src/jvmMain/kotlin/okhttp3/AsyncDns.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package okhttp3
+
+import java.net.InetAddress
+import java.net.UnknownHostException
+import java.util.concurrent.CountDownLatch
+import okio.IOException
+
+/**
+ * An async domain name service that resolves IP addresses for host names.
+ *
+ * The main implementations will typically be implemented using specific Dns libraries such as
+ *  * Android DnsResolver
+ *  * OkHttp DnsOverHttps
+ *  * dnsjava Resolver
+ *
+ * Implementations of this interface must be safe for concurrent use.
+ */
+interface AsyncDns {
+  /**
+   * Query dns records for `hostname`, in the order they are received.
+   */
+  fun query(hostname: String, callback: Callback)
+
+  /**
+   * Callback to receive results from the Dns Queries.
+   */
+  interface Callback {
+    /**
+     * Return addresses for a dns question,
+     * typically but always a single class of IPv4 (A) or IPv6 (AAAA).
+     * May be empty.
+     */
+    fun onResponse(hostname: String, addresses: List<InetAddress>)
+
+    /**
+     * Returns an error for the dns query.
+     */
+    fun onFailure(hostname: String, e: IOException)
+  }
+
+  /**
+   * Class of Dns addresses, such that clients that treat these differently, such
+   * as attempting IPv6 first, can make such decisions.
+   */
+  enum class DnsClass(val type: Int) {
```

Comment:
Need this be a public API? (I think it has to be??)

Reply:
Yep - part of public API.